### DISCUSSION TITLE
[solved] read serial data

### DISCUSSION THREAD

**Discussion by git5any** [arendst/Tasmota#15776] [Q&A] [answered]
Hi,
I have a Leveljet/ProJet Device to get the volume of my cistern. It has a serial output, defined 19200,8,N,1, data is sent every second. The output are 12 bytes:
0  Gerätekennung Low-Byte $E8 
1  Gerätekennung High-Byte $03
2  Distanz Low-Byte
3  Distanz High-Byte
4  Füllhöhe Low-Byte
5  Füllhöhe High-Byte
6  Liter Low-Byte
7  Liter High-Byte
8  Inhalt in Prozent
9  Zustand derAusgänge
10  Kontrollbyte Low-Byte
11  Kontrollbyte High-Byte

The value I need is in byte 8.
Since years I use an arduino with mysensors to read this serial and send it to my homeautomation. Now I try to get rid of the "wireless"-stuff and use a WT32-ETH01 (a ESP32 with Ethernet) connected to my LAN.
The ESP is customflashed with Ethernet, scripting and the serial-stuff and works.
What I tried in the script so far:

```
>D
in=""
>B
so(5 -1 19200 8N1)
>S
if upsecs%10==0 {
    in=sr()
    byt=sa()
    print in: %in%  bytes: %0byt%
}
```
produces this:
```
11:51:32.083 Script: nv=3, tv=1, vns=16, vmem=176, smem=8192
11:51:32.215 Serial port set to 8N1 19200 bit/s at rx=5 tx=-1 rbu=128
11:51:33.026 in:   bytes: 11()???
11:51:43.020 in:   bytes: 128()???
11:51:53.020 in: e  bytes: 246()???
11:52:03.037 in: >
11:52:13.044 in:   bytes: 251()???
11:52:23.050 in:   bytes: 248()???
11:52:33.031 in: e  bytes: 246()???
11:52:43.021 in: >
11:52:53.033 in:   bytes: 239()???
11:53:03.015 in:   bytes: 236()???
11:53:13.041 in: e  bytes: 234()???
11:53:23.021 in: >
```
when I try in=srb() its like this:
```
11:54:33.663 Serial port set to 8N1 19200 bit/s at rx=5 tx=-1 rbu=128
11:54:43.030 in: ���?  bytes: 119()???
11:54:53.039 in:   bytes: 238()???
11:55:03.034 in:   bytes: 249()???
11:55:13.041 in:   bytes: 248()???
11:55:23.041 in:   bytes: 247()???
11:55:33.063 in:   bytes: 246()???
```
and with in=srb() ist this:
```
11:56:26.527 Serial port set to 8N1 19200 bit/s at rx=5 tx=-1 rbu=128
11:56:33.049 in:   bytes: 72()???
11:56:43.031 in:   bytes: 192()???
11:56:53.036 in: ���?  bytes: 252()???
11:57:03.027 in:   bytes: 252()???
11:57:13.052 in:   bytes: 252()???
11:57:23.047 in:   bytes: 252()???
11:57:33.053 in:   bytes: 252()???
```

It looks like that there is data coming in, but I'm not able to read it correctly. I think I just dont undrestand the description, and if it is possible that way at all.
On the Arduino I read in a loop from (soft)serial into a array until I read 12 bytes and the I just use 8th byte which contains the cistern-level in percent.

From the hardwareside the WT32-ETH01 has 2 Serial ports on pins, one is labeld "reserved for flashing" with RXD0 on GPIO03, TXD0 on GPIO01 and the other is labeled  RXD2 on GPIO05 and TXD2 on GPIO17. Therefor I use GPIO05, with an 1.8k/3.3k resistor-divider for the signalline that delivers 5v pulses.

Do I also need some looping? Is >S the right section? 

Thanks for any help.

**Comment by barbudor** [comment]
As you have an ESP32, and if you are more familiar with standard programming languages you may want to give a try to Berry which is a python like scripting language available with tasmota32

Not saying that you can't achieve that with scripting, but I've always been too lazy to really take the time to take the learning step.

**Comment by gemu2015** [comment]
check here where i gave some examples how to decode serial data
https://github.com/arendst/Tasmota/discussions/15696

this is a corrected version of your script above, if you wait 10 seconds between queries you will get to much data, so check every 2 seconds or less, but i guess it would be better to read arrays like in the link above

\>D
in=""
byt=0
\>B
so(5 -1 19200 8N1)
\>S
if upsecs%2==0 {
; check if bytes available 
byt=sa()
if byt>0 {
in=sr()
print in: %in%  bytes: %0byt%
}
}

**Comment by gemu2015** [comment]
array version
\>D
m:a_in=0 12
byt=0
\>B
so(5 -1 19200 8N1)
\>S
if upsecs%2==0 {
; check if bytes available
byt=sa()
if byt>=12 {
sra(a_in)
if a_in[1]==0xe8 {
print in: %a_in[9]% bytes: %0byt%
}
}
}

**Comment by git5any** [comment]
Hi,
short Version: it works!
long: Details tomorrow, must cancel my investigation, sorry.

**Comment by git5any** [comment, marked as answer]
Hi again,
now the long version. With the information from the above mentioned link I was able to find out that I have to use an arry to get the data. In the same moment the update from gemu2015 came in and i gave it a try. As it didn't work directly, I removed the check for the first byte and then the data was accessible.
```
>D
;create the input-array
m:a_in=0 12
;number of bytes
byt=0

>B
;open the serial port RX on GPIO5, no TX, 19200bd, 8N1
so(5 -1 19200 8N1)

>S
; check every secon if bytes available
byt=sa()
;bytes avail (at least 12)
if byt>=12 {
	;reset some data in the array
	a_in[1]=-99
	a_in[2]=-99
	a_in[9]=-99
	;fill the array from serial
	sra(a_in)
	;check the first 2 items in array
	if a_in[1]==0 
	and a_in[2]==16 {
		;print out all items in the array
		print Type: %0a_in[1]% %0a_in[2]% dist: %0a_in[3]% %0a_in[4]% Hight: %0a_in[5]% %0a_in[6]% Liter: %0a_in[7]% %0a_in[8]% Proz: %0a_in[9]% Out: %0a_in[10]%  check: %0a_in[11]% %0a_in[12]% (bytes: %0byt%)
		;reset byte-variable to 0
		byt=0
	}
}


```
the output:
11:38:54.746 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:38:55.746 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:38:56.751 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:38:57.750 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:38:58.795 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:38:59.769 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:39:00.770 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
11:39:01.771 Type: 0 16 dist: 19 0 Hight: 101 0 Liter: 62 1 Proz: 100 Out: 0  check: 23 14 (bytes: 12)
  Now I check for if the first two values in the array are 0 and 16, because they do not change and then I'm quite certain to get the right number in the 8th position of the array, the percent-Value.
  It also shows the numbers are already in a readable form, so I can use the percent-value as it is.
  
  Once more thank you very much for all the information you provided here to solve this.
  
  I will also have a look into Berry-scripting, but I just found out the "normal"-scripting could be usefull and had some success for other parts of my goal: final stage is an ethernt-coupled ESP32 reading cistern via serial, SML-Meter via script, a S0-counter via script and the 4 indicator-leds from another device. Data sent via MQTT to fhem. All in one box, all in one room. The single parts now work, will try to fit them together.
